applications/TrilinosApplication/custom_strategies/builder_and_solvers/trilinos_block_builder_and_solver.h

```diff
@@ -817,12 +864,108 @@ class TrilinosBlockBuilderAndSolver
         KRATOS_CATCH("");
     }
 
+    /**
+     * @brief Applies the constraints with master-slave relation matrix (RHS only)
+     * @param pScheme The integration scheme considered
+     * @param rModelPart The model part of the problem to solve
+     * @param rb The RHS vector
+     */
+    void ApplyRHSConstraints(
+        typename TSchemeType::Pointer pScheme,
+        ModelPart& rModelPart,
+        TSystemVectorType& rb
+        ) override
+    {
+        KRATOS_TRY
+
+        if (rModelPart.GetCommunicator().GlobalNumberOfMasterSlaveConstraints() > 0) {
+            BuildMasterSlaveConstraints(rModelPart);
+
+            // Reference to T
+            const TSystemMatrixType& r_T = *mpT;
+
+            // Compute T' b
+            const TSystemVectorType copy_b(rb);
+            TSparseSpace::TransposeMult(r_T, copy_b, rb);
+
+            // Apply diagonal values on slaves
+            IndexPartition<std::size_t>(mSlaveIds.size()).for_each([&](std::size_t Index){
+                const IndexType slave_equation_id = mSlaveIds[Index];
+                if (mInactiveSlaveDofs.find(slave_equation_id) == mInactiveSlaveDofs.end()) {
+                    TrilinosAssemblingUtilities::SetGlobalValueWithoutGlobalAssembly(rb, slave_equation_id, 0.0);
+                }
+            });
+
+            // Global assembly
+            rb.GlobalAssemble();
+        }
+
+        KRATOS_CATCH("")
+    }
+
+    /**
+     * @brief Applies the constraints with master-slave relation matrix
+     * @param pScheme The integration scheme considered
+     * @param rModelPart The model part of the problem to solve
+     * @param rA The LHS matrix
+     * @param rb The RHS vector
+     */
+    void ApplyConstraints(
+        typename TSchemeType::Pointer pScheme,
+        ModelPart& rModelPart,
+        TSystemMatrixType& rA,
+        TSystemVectorType& rb
+        ) override
+    {
+        KRATOS_TRY
+
+        if (rModelPart.GetCommunicator().GlobalNumberOfMasterSlaveConstraints() > 0) {
+            BuildMasterSlaveConstraints(rModelPart);
+
+            // Reference to T
+            const TSystemMatrixType& r_T = *mpT;
+
+            // Compute T' A T
+            const TSystemMatrixType copy_A(rA);
```

Comment:
In fact if you look at the details of the implementation you will see that there are some tricks in order to avoid problems with Trilinos, this is because I found a bug in Epetra that a priori Trilinos is not going to fix as Epetra is already legacy.

Reply:
I assure it works, I did many checks

Reply:
I will resolve the conversation, if you have more questions we can speak directly.